pkg/auth/ldaputil/query.go

```diff
@@ -0,0 +1,211 @@
+	}
+}
+
+// buildDNQuery builds the query that finds an LDAP entry with the given DN
+// this is done by setting the DN to be the base DN for the search and setting the search scope
+// to only consider the base object found
+func (o *IdentifiyingLDAPQueryOptions) buildDNQuery(dn string,
+	attributes []string) *ldap.SearchRequest {
+	return ldap.NewSearchRequest(
+		dn,
+		ldap.ScopeBaseObject, // over-ride original
+		int(o.DerefAliases),
+		0, // allowed return size - indicates no limit
+		o.TimeLimit,
+		false,           // not types only
+		"objectClass=*", // filter that returns all values
+		attributes,
+		nil, // no controls
+	)
+}
+
+// buildAttributeQuery builds the query containing a filter that conjoins the common filter given
+// in the configuration with the specific attribute filter for which the attribute value is given
+func (o *IdentifiyingLDAPQueryOptions) buildAttributeQuery(attributeValue string,
+	attributes []string) *ldap.SearchRequest {
+	specificFilter := fmt.Sprintf("%s=%s",
+		ldap.EscapeFilter(o.QueryAttribute),
+		ldap.EscapeFilter(attributeValue))
+
+	filter := fmt.Sprintf("(&(%s)(%s))", o.Filter, specificFilter)
+
+	return ldap.NewSearchRequest(
+		o.BaseDN,
+		int(o.Scope),
+		int(o.DerefAliases),
+		0, // allowed return size - indicates no limit
+		o.TimeLimit,
+		false, // not types only
+		filter,
+		attributes,
+		nil, // no controls
+	)
+}
+
+// QueryForUniqueEntry queries for an LDAP entry on an LDAP server determined from a ClientConfig
+// by creating a search request from the requisite input. The query is expected to return one unqiue
+// result. If this is not the case, errors are raised
+func QueryForUniqueEntry(clientConfig LDAPClientConfig,
+	query *ldap.SearchRequest) (entry *ldap.Entry, err error) {
+
+	result, err := QueryForEntries(clientConfig, query)
+	if err != nil {
+		return nil, err
+	}
+	if len(result) > 1 {
+		return nil, fmt.Errorf("multiple entries found matching %s", query.Filter)
+	}
+	entry = result[0]
+	glog.V(4).Infof("found dn=%q for %s", entry.DN, query.Filter)
+	return entry, nil
+}
+
+// QueryForEntries queries for LDAP entries on an LDAP server determined from a ClientConfig by
+// creating a search request from the requisite input.
+func QueryForEntries(clientConfig LDAPClientConfig,
+	query *ldap.SearchRequest) (result []*ldap.Entry, err error) {
+
+	connection, err := clientConfig.Connect()
+	if err != nil {
+		return nil, err
+	}
+	defer connection.Close()
+
+	glog.V(4).Infof("searching LDAP server for %s", query.Filter)
+	searchResult, err := connection.Search(query)
+	if err != nil {
+		return nil, err
+	}
+
+	entries := searchResult.Entries
+	// No entries returned from the LDAP search request means that the LDAP search was not configured
+	// correctly. The search must return with at least one LDAP entry
+	if len(entries) == 0 {
+		return nil, fmt.Errorf("no LDAP user entry found for filter: %s", query.Filter)
```

Comment:
> I believe the reason that this errors on empty is that when you are looking for a specific LDAP entry that you know exists (as you have the UID for it), the behavior of the LDAP sync would not be as expected if the entry was not found. For example, if you want to sync a group and screwed up the user filter such that the users of the group are not a subset of all users from that filter, the sync will effectively do nothing, and you'd expect as a user to get an error instead of silently doing nothing.

Find usages and fix it.  A `List` that returns `{}` did not fail.